vendors/st/boards/stm32l475_discovery/ports/common_io/iot_i2c.c

```diff
@@ -0,0 +1,744 @@
+    HAL_NVIC_SetPriority( DISCOVERY_I2Cx_EV_IRQn, 0x0F, 0 );
+    HAL_NVIC_EnableIRQ( DISCOVERY_I2Cx_EV_IRQn );
+}
+
+void HAL_I2C_MspDeInit( I2C_HandleTypeDef * hi2c )
+{
+    GPIO_InitTypeDef xGpio_init_structure;
+
+    /* Configure I2C Tx, Rx as alternate function */
+    xGpio_init_structure.Pin = DISCOVERY_I2Cx_SCL_PIN | DISCOVERY_I2Cx_SDA_PIN;
+    HAL_GPIO_DeInit( DISCOVERY_I2Cx_SCL_SDA_GPIO_PORT, xGpio_init_structure.Pin );
+
+    DISCOVERY_I2Cx_SCL_SDA_GPIO_CLK_DISABLE();
+
+    DISCOVERY_I2Cx_CLK_DISABLE();
+}
+
+void HAL_I2C_MasterTxCpltCallback( I2C_HandleTypeDef * hi2c )
+{
+    IotI2CDescriptor_t * pDescriptor = _toDescriptor( hi2c );
+    BaseType_t higherPriorityTaskWoken = pdFALSE;
+
+    /* Subtract not transferred yet. */
+    pDescriptor->usTransmittedTxBytes -= hi2c->XferCount;
+
+    if( pDescriptor->xCallback != NULL )
+    {
+        pDescriptor->xCallback( eI2CCompleted, pDescriptor->pvUserContext );
+    }
+
+    xSemaphoreGiveFromISR( pDescriptor->xSemphr, &higherPriorityTaskWoken );
+    portYIELD_FROM_ISR( higherPriorityTaskWoken );
+}
+
+void HAL_I2C_MasterRxCpltCallback( I2C_HandleTypeDef * hi2c )
+{
+    IotI2CDescriptor_t * pDescriptor = _toDescriptor( hi2c );
+    BaseType_t higherPriorityTaskWoken = pdFALSE;
+
+    /* Subtract not transferred yet. */
+    pDescriptor->usReceivedRxBytes -= hi2c->XferCount;
+
+    if( pDescriptor->xCallback != NULL )
+    {
+        pDescriptor->xCallback( eI2CCompleted, pDescriptor->pvUserContext );
+    }
+
+    xSemaphoreGiveFromISR( pDescriptor->xSemphr, &higherPriorityTaskWoken );
+    portYIELD_FROM_ISR( higherPriorityTaskWoken );
+}
+
+void HAL_I2C_ErrorCallback( I2C_HandleTypeDef * hi2c )
+{
+    IotI2CDescriptor_t * pDescriptor = _toDescriptor( hi2c );
+    BaseType_t higherPriorityTaskWoken = pdFALSE;
+
+    if( pDescriptor->xCallback != NULL )
+    {
+        pDescriptor->xCallback( eI2CCompleted, pDescriptor->pvUserContext );
+    }
+
+    xSemaphoreGiveFromISR( pDescriptor->xSemphr, &higherPriorityTaskWoken );
+    portYIELD_FROM_ISR( higherPriorityTaskWoken );
+}
+
+/*---------------------Private functions---------------------*/
+
+static int32_t _toI2cStatus( HAL_StatusTypeDef halStatus,
+                             uint8_t operation )
+{
+    switch( halStatus )
+    {
+        case HAL_OK:
+            return IOT_I2C_SUCCESS;
+
+        case HAL_ERROR:
+            return operation == _I2C_READ_OP ? IOT_I2C_READ_FAILED : IOT_I2C_WRITE_FAILED;
+
+        case HAL_BUSY:
+            return IOT_I2C_BUSY;
+
+        case HAL_TIMEOUT:
+            return IOT_I2C_BUS_TIMEOUT;
+    }
+
+    /* TODO: it is better to assert(0) here. */
+    return IOT_I2C_SUCCESS;
+}
+
+static int32_t _doMasterTransfer( IotI2CDescriptor_t * pI2cDescriptor,
+                                  uint8_t * const pucBuffer,
+                                  size_t xBytes,
+                                  HAL_StatusTypeDef ( * pMasterSync )( I2C_HandleTypeDef * hi2c,
```

Comment:
Why aren't these function prototypes based on shared definitions?

Reply:
This is defined by ST HAL, not controlled by us.

Reply:
Some other piece of code presumably has implementations of these routines. Wouldn't these types need to be visible there, too?

Reply:
Not quite understand your question. One of the implementation is this https://github.com/aws/amazon-freertos/blob/master/vendors/st/stm32l475_discovery/STM32L4xx_HAL_Driver/stm32l4xx_hal_i2c.c#L753